### ISSUE TITLE
Random errors with redefinition of template and no such template

### ISSUE THREAD

**Issue by jarlah** [jdbi/jdbi#1052]
Update:
Found a way to reproduce it consistently. See https://github.com/jarlah/jdbi-stringtemplate4-racecondition-bug

Hi

I am trying to use Jdbi with stringtemplate4. I have created two template files that imports another template file relatively. e.g. import "../whatever/WhateverDao.sql.stg".

The problem occurs randomly, and I think is also solved randomly, that the methods in WhateverDao.sql.stg is not found or being overwritten (nulled out) by thread synchronization issues.

I am using:
spring boot 2.0.0.RELEASE
Jdbi 3.1.0

I have tried to create a JdbiFactoryBean and a normal Jdbi creation. 

I get red errors in console when the random error occurs saying that the template method is not found. 
Below I see other red errors saying the methods have been redefined.
I suppose those two errors are linked.

But I also think that there is a major issue with synchronization.

Does jdbi support import statements in template group files? If not then templates are uselees?

**Comment by jarlah** [comment]
I will get back with concrete error messages but it would be great to know if anyone has tried to import another template  etc and if it's supported. Jdbi doc dont say no but not yes either.

**Comment by arteam** [comment]
Looks like it's relevant to #971/#973.

**Comment by jarlah** [comment]
@arteam it might be related. But I have a hard time figuring out when it was merged into master.
Anyway without giving away too much information:
```
context [searchLegacySessions] 5:88 no such template: delayedCondition
context [searchLegacySessions] 5:89 passed 1 arg(s) to template null with 0 declared arg(s)
context [searchLegacySessions] 6:90 no such template: isLabourCompleted
context [searchLegacySessions] 7:92 no such template: isLabourOverlapping
context [searchLegacySessions] 5:88 no such template: delayedCondition
context [searchLegacySessions] 5:89 passed 1 arg(s) to template null with 0 declared arg(s)
context [searchLegacySessions] 6:90 no such template: isLabourCompleted
context [searchLegacySessions] 7:92 no such template: isLabourOverlapping
context [searchLegacySessions] 5:88 no such template: delayedCondition
context [searchLegacySessions] 5:89 passed 1 arg(s) to template null with 0 declared arg(s)
context [searchLegacySessions] 6:90 no such template: isLabourCompleted
context [searchLegacySessions] 7:92 no such template: isLabourOverlapping
LabourDao.sql.stg 85:0: redefinition of template isLabourOverlapping
LabourDao.sql.stg 120:0: redefinition of template delayedCondition
LabourDao.sql.stg 128:0: redefinition of template isLabourCompleted
context [searchLegacySessions] 5:89 passed 1 arg(s) to template null with 0 declared arg(s)
context [searchLegacySessions] 6:90 no such template: isLabourCompleted
context [searchLegacySessions] 7:92 no such template: isLabourOverlapping
```
Other than the fact that I havent set default params everywhere, you can see that the LabourDao.sql.stg file is suddenly redefined, possibly by the import.

**Comment by jarlah** [comment]
Stupid me why do i need imports. I can just load a template method programmatically in my declarative default method and just pass it in as a defined string;) problem solved and it's actually not that hacky either. It's actually cleaner.

**Comment by qualidafial** [comment]
Per the `RELEASE_NOTES` file, cross-group imports was added in 3.1.0

**Comment by jarlah** [comment]
Well it does not work with my spring boot 2 setup. Whatsoever.

**Comment by jarlah** [comment]
I am in the prosess of making a test bench project where my plan is to fire up two threads that calls to different endpoints with different daos but using the same template file under the hood. Or where one of the template files for one of the daos is imported by the other

**Comment by jarlah** [comment]
https://github.com/jarlah/jdbi-stringtemplate4-racecondition-bug

**Comment by jarlah** [comment]
Take that ;) Finally, i reproduced it constantly in a unit test :dancer:

**Comment by jarlah** [comment]
@qualidafial :100: reproducable

**Comment by jarlah** [comment]
The bug is actually about two threads callling the same dao. So i have had many wrong conclusions along the way

**Comment by jarlah** [comment]
Can someone confirm that this is an actual bug? Is it possible to preload all templates at startup maybe? Or disable cache? Manually triggering template loading at startup is a fragile solution because it's template specific.

**Comment by qualidafial** [comment]
Hi @jarlah, thanks for chasing this down. Give us a chance to dig into your test and let's see what we can figure out.

**Comment by qualidafial** [comment]
Hi @jarlah, sorry for the slow response. I've been pretty busy the past few weeks, but I'll try to take a look at this tonight.

**Comment by stevenschlansker** [comment]
Thanks for the test case, that highlighted the bug.  (although the test case has a race of its own!)

**Comment by jarlah** [comment]
@stevenschlansker When will the fix be released? Can you cherrypick the fix into a 3.1.2 release?

**Comment by qualidafial** [comment]
The next release will be 3.2.0 (new features added), and we could cut another release tonight. I'll try to set aside time for that.
